### DISCUSSION TITLE
3D Mouse Support?

### DISCUSSION THREAD

**Discussion by SteveAtSWS** [godotengine/godot-proposals#9240] [3D]
If there aren't already plans for support for 3D Mouse devices (like [3DConnexion's Space Mouse](https://3dconnexion.com/uk/spacemouse/)), I'd like to put it on the godot dev's radar.

This would be an indispensable feature for usability and ease of access.

**Comment by tebz** [comment]
Or any joystick ?
If exposing 
  Node3DEditorPlugin::get_state()
and
  Node3DEditorPlugin::set_state()
for the cameras to GdScript EditorPlugin, i think it might solve the problem.
Would make it possible to make editor script that supports any 3D input device, as well as navigating using a Joystick
"3DConnexion's Space Mouse" is just a 6 axis Joystick.
Also - 
  To get a list of the 3D cameras and camera controls in a clean way would be nicer than to scan the tree for them. 
  Or perhaps just the currently active camera/viewport together with set_state() and get_state()
.
And it would also be possible to save and restore camera positions so we can navigate quickly in larger designs.    
.
Of course, there are more ways to do this but the functions are:
* Move, 6 degrees of freedom, on one (the perspective) or the active 3D camera with a EditorPlugin script.
* Save and restore camera position and rotation. Perhaps with a EditorPlugin script.
.
A lot of 3D applications have this features. Like Blender and a lot of other applications.
.
I have made an GdScript EditorPlugin that moves the active editor 3Dcamera, but as soon as you "_update_camera()" the camera, for example try to select something, the camera position are reset to the cursor values. Just moving the camera does not fully work.
